### PR TITLE
#260 update mutations that affect price

### LINKED ISSUE #260: Update mutations that affect price
Continuation of #214

### PR DESCRIPTION
**Pull request by roxy-dao**
Closes #260 

F/e changes coming up --- 

- Changed most total before tax to be optional, and calculate the total before tax if it is not provided.

### CODE REVIEW

File: server/graphql/invoice_line/src/mutations/inbound_shipment_line/line/insert.rs [hunk 1]
```diff
@@ -26,8 +26,8 @@ pub struct InsertInput {
     pub sell_price_per_pack: f64,
     pub expiry_date: Option<NaiveDate>,
     pub number_of_packs: u32,
-    pub total_before_tax: f64,
-    pub tax: Option<TaxUpdate>,
+    pub total_before_tax: Option<PriceInput>,
+    pub tax: Option<PriceInput>,
```
Comment (clemens-msupply) [inline server/graphql/invoice_line/src/mutations/inbound_shipment_line/line/insert.rs:30]: Why not have TaxInput, and TotalBeforeTaxInput? seems a bit confusing if you just look at the api, e.g. why be able to set a percentage on total_before_tax if it is then silently ignored.

File: server/service/src/invoice_line/inbound_shipment_service_line/update/mod.rs [hunk 2]
```diff
@@ -5,18 +5,14 @@ use generate::generate;
...
-};
+use crate::{invoice_line::query::get_invoice_line, service_provider::ServiceContext, WithDBError};
 #[derive(Debug, Clone, PartialEq, Default)]
 pub struct UpdateInboundShipmentServiceLine {
     pub id: String,
     pub item_id: Option<String>,
     pub name: Option<String>,
     pub total_before_tax: Option<f64>,
-    pub tax: Option<ShipmentTaxUpdate>,
+    pub tax: Option<f64>,
```
Comment (clemens-msupply) [inline server/service/src/invoice_line/inbound_shipment_service_line/update/mod.rs:15]: How to unset a tax value? With ShipmentTaxUpdate you where able to:
1) not touch the existing tax value in the update using `tax = None`
2) unset the tax value with `tax = Some(ShipmentTaxUpdate { percentage: None})`
3) update the tax value with `tax = Some(ShipmentTaxUpdate { percentage: Some(5.0)})`

File: server/graphql/invoice_line/src/mutations/inbound_shipment_line/line/insert.rs [hunk 3]
```diff
@@ -26,8 +27,8 @@ pub struct InsertInput {
     pub sell_price_per_pack: f64,
     pub expiry_date: Option<NaiveDate>,
     pub number_of_packs: u32,
-    pub total_before_tax: f64,
-    pub tax: Option<TaxUpdate>,
+    pub total_before_tax: Option<PriceInput>,
```
Comment (mark-prins) [inline server/graphql/invoice_line/src/mutations/inbound_shipment_line/line/insert.rs:30]: the struct has a `total_before_tax` property, so we'll end up with `total_before_tax.total_before_tax` won't we? 😕 
Why not just a float? or have a `pub price` property here, with all the tax bits in it?

File: server/graphql/invoice_line/src/mutations/inbound_shipment_line/line/insert.rs [hunk 4]
```diff
@@ -103,8 +104,13 @@ impl InsertInput {
             sell_price_per_pack,
             cost_price_per_pack,
             number_of_packs,
-            total_before_tax,
-            tax: tax.and_then(|tax| tax.percentage),
+            total_before_tax: total_before_tax
+                .and_then(|total_before_tax| total_before_tax.total_before_tax),
```
Comment (mark-prins) [inline server/graphql/invoice_line/src/mutations/inbound_shipment_line/line/insert.rs:108]: ah. there it is.

**Review by mark-prins — COMMENTED** [review]
I'm slightly confused - so should probably discuss. or just walk away and leave to you & Clemens 😁

File: server/service/src/invoice_line/inbound_shipment_line/insert/mod.rs [hunk 5]
```diff
@@ -23,8 +27,8 @@ pub struct InsertInboundShipmentLine {
     pub sell_price_per_pack: f64,
     pub expiry_date: Option<NaiveDate>,
     pub number_of_packs: u32,
-    pub total_before_tax: f64,
-    pub tax: Option<f64>,
+    pub total_before_tax: Option<f64>,
+    pub tax: Option<ShipmentTaxUpdate>,
```
Comment (andreievg) [inline server/service/src/invoice_line/inbound_shipment_line/insert/mod.rs:31]: I think it makes sense for insert to have just Option<f64> (there is no need to 'unset' it for inserts)

File: server/service/src/invoice_line/inbound_shipment_line/insert/generate.rs [hunk 6]
```diff
@@ -68,9 +68,21 @@ fn generate_line(
         item_name,
         item_code,
         stock_line_id: None,
-        total_before_tax,
-        total_after_tax: total_after_tax(total_before_tax, tax),
-        tax,
+        total_before_tax: match total_before_tax {
```
Comment (andreievg) [inline server/service/src/invoice_line/inbound_shipment_line/insert/generate.rs:71]: would prefer these to be done above in a variable, since you doing the same thing twice (it's easier to read if you just do it once at the top and less chance of error), i.e. when setting total_after_tax it would become:
total_after_tax: total_after_tax(total_before_tax)

File: server/service/src/invoice_line/inbound_shipment_line/update/generate.rs [hunk 7]
```diff
@@ -90,6 +93,18 @@ fn generate_line(
...
+    if let Some(total_before_tax) = total_before_tax {
+        update_line.total_before_tax = total_before_tax;
+    } else {
+        update_line.total_before_tax = cost_price_per_pack
+            .unwrap_or(update_line.cost_price_per_pack)
+            * number_of_packs.unwrap_or(i32_to_u32(update_line.number_of_packs)) as f64;
+    }
+
+    if let Some(tax) = tax {
+        update_line.tax = tax.percentage;
```
Comment (andreievg) [inline server/service/src/invoice_line/inbound_shipment_line/update/generate.rs:105]: can use unwrap_or (as per the rest of the method)

**Review by andreievg — APPROVED** [review]
Look pretty good, some stylistic comments.
I question a couple of things:
* I wonder why we need `PriceInput` (we don't need to 'unset' the total_before_tax). 
* Don't really need TaxInput for insert in service
I.e. for above it's probably good to have a generic inputs for graphql, but not for service.

Also I think I might have mislead you in one of our conversation (or at least didn't quite think it through), you can do this change in another PR btw:
For updates to `non service` lines, the logic should be:
```
if total_before_tax is not null us it 
if total before tax is null:
(for both)
if number of packs is change, recalculate total_before_tax
(for inbound)
if cost price per pack is changed, recalculate total_before_tax
(for oubound)
if stock line id or item_id is change, recalculate total_before_tax
```

**Comment by roxy-dao** [comment]
> Look pretty good, some stylistic comments. I question a couple of things:
> 
> * I wonder why we need `PriceInput` (we don't need to 'unset' the total_before_tax).
> * Don't really need TaxInput for insert in service
>   I.e. for above it's probably good to have a generic inputs for graphql, but not for service.
> 
> Also I think I might have mislead you in one of our conversation (or at least didn't quite think it through), you can do this change in another PR btw: For updates to `non service` lines, the logic should be:
> 
> ```
> if total_before_tax is not null us it 
> if total before tax is null:
> (for both)
> if number of packs is change, recalculate total_before_tax
> (for inbound)
> if cost price per pack is changed, recalculate total_before_tax
> (for oubound)
> if stock line id or item_id is change, recalculate total_before_tax
> ```

if total_before_tax is not null us it 
^ What does us it mean? 🤔

File: server/service/src/invoice_line/inbound_shipment_line/insert/generate.rs [hunk 8]
```diff
@@ -68,9 +68,21 @@ fn generate_line(
...
+            total_before_tax.unwrap_or(cost_price_per_pack * number_of_packs as f64),
+            match tax.clone() {
+                Some(tax) => tax.percentage,
+                None => Some(0.0),
+            },
+        ),
+        tax: match tax {
+            Some(tax) => tax.percentage,
+            None => Some(0.0),
+        },
```
Comment (clemens-msupply) [inline server/service/src/invoice_line/inbound_shipment_line/insert/generate.rs:85]: if tax is not set it should be set to None not to Some(0.0)

File: server/service/src/invoice_line/inbound_shipment_service_line/insert/generate.rs [hunk 9]
```diff
@@ -20,8 +20,17 @@ pub fn generate(
...
+            total_before_tax,
+            match tax.clone() {
+                Some(tax) => tax.percentage,
+                None => Some(0.0),
+            },
+        ),
+        tax: match tax {
+            Some(tax) => tax.percentage,
+            None => Some(0.0),
+        },
```
Comment (clemens-msupply) [inline server/service/src/invoice_line/inbound_shipment_service_line/insert/generate.rs:33]: Same here don't set the tax to 0.0 if not specified

File: server/service/src/invoice_line/outbound_shipment_line/insert/generate.rs [hunk 10]
```diff
@@ -77,9 +77,21 @@ fn generate_line(
...
+        total_after_tax: total_after_tax(
+            total_before_tax.unwrap_or(cost_price_per_pack * number_of_packs as f64),
+            match tax.clone() {
+                Some(tax) => tax.percentage,
+                None => Some(0.0),
+            },
+        ),
+        tax: match tax {
+            Some(tax) => tax.percentage,
+            None => Some(0.0),
```
Comment (clemens-msupply) [inline server/service/src/invoice_line/outbound_shipment_line/insert/generate.rs:93]: None => None

**Comment by andreievg** [comment]
> if total_before_tax is not null us it
^ What does us it mean? 🤔

not `None` oops (since it's currently nested in PriceInput, need to test for the nested `None`, but if it was just Option<f64>, just for that one being `None`). The reason I said `null` is because it's currently nested, and that's the only way we can set Optional nullable in graphql to null, which is technically not needed for price, but it's needed for tax

**Review by clemens-msupply — COMMENTED** [review]
See comments, when inserting a new line and tax is not specified we should not set the tax on the line...

**Comment by roxy-dao** [comment]
> [34499f9](https://github.com/openmsupply/open-msupply/pull/306/commits/34499f9e44bd8a627cb802a0746a1672607c7458) Will do the logic in another PR~
